code/game/objects/items/tools/multifunctiontool.dm

```diff
@@ -1,76 +1,89 @@
-//Power Tools
-/obj/item/powertool
-	var/stateswap = 0 //what state is it in
-	var/icon_state_swap = 0 //which icon should it be
-	var/tool_behaviour_swap = 0 //what tool behaviour(s) should be here
-	var/swapsound = 0 //the sound that plays for swapping
+//Multifunction tools
+/obj/item/multifunctiontool
+	var/list/tool_icon_states = list() //which icon should it be
+	var/list/tool_behaviours = list()  //what tool behaviour(s) should be here
+	var/list/tool_mode_names = list() //text added to the examine description
+	var/active_behaviour_index = 1
+	var/swapsound //the sound that plays for swapping
 
-/obj/item/powertool/attack_self(mob/living/carbon/user) //activating the item causes it to swap icons and behaviours
-	stateswap = !stateswap
-	if(stateswap)
-		icon_state = icon_state_swap
-		tool_behaviour = tool_behaviour_swap
-	else
-		icon_state = initial(icon_state)
-		tool_behaviour = initial(tool_behaviour)
+/obj/item/multifunctiontool/attack_self(mob/living/carbon/user) //activating the item causes it to swap icons and behaviours
+	if(tool_behaviours.len > 1)
+		select_mode(user)
+	tool_behaviour = tool_behaviours[active_behaviour_index]
+	icon_state = tool_icon_states[active_behaviour_index]
 	playsound(get_turf(user), swapsound, 50, 1)
-	to_chat(user, "<span class='notice'>You attach a different bit to [src].</span>")
+	to_chat(user, "<span class='notice'>You use a different part on [src].</span>")
+
+/obj/item/multifunctiontool/proc/select_mode(mob/living/user)
```

Comment:
Several issues here (read: probably just one issue, but can be addressed in multiple different ways):
1) I think a better name would be `cycle_mode`. I think `select` implies that it gets passed an index of the mode you want to switch towards.
2) this function doesn't actually change the mode, it only adjusts the index. It would be cleaner if it *returned* the  desired active_behaviour_index value instead of adjusting it in-place
3) As is, i would much prefer this to be inlined into attack_self, and not be a standalone function
4) It can be condensed from 3 lines down to one `active_behaviour_index = (active_behaviour_index + 1) % tool_behaviours.len + 1`

I think at least either 2 or 3 need to be addressed here (the rest are whatever-ish)